crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -0,0 +1,327 @@
+                    .clone()
+                    .into_iter()
+                    .peekable()
+                    .skip_while(|attr| attr.as_ref() == ""),
+            );
+            attrpath
+        };
+
+        let nixpkgs_flakeref = FlakeRef::Indirect(IndirectRef::new(
+            format!("nixpkgs-{}", self.stability),
+            Default::default(),
+        ));
+
+        let analyzer_flakeref = FlakeRef::Path(PathRef::new(
+            PathBuf::from(env!("FLOX_ANALYZER_SRC")),
+            Default::default(),
+        ));
+
+        let eval_analysis_command = Eval {
+            flake: FlakeArgs {
+                override_inputs: [
+                    ("target".to_string(), self.publish_ref.clone().into_inner()).into(),
+                    (
+                        "target/flox-floxpkgs/nixpkgs/nixpkgs".to_string(),
+                        nixpkgs_flakeref,
+                    )
+                        .into(),
+                ]
+                .to_vec(),
+                no_write_lock_file: true.into(),
+            },
+            eval_args: runix::arguments::EvalArgs {
+                installable: Some(
+                    Installable {
+                        flakeref: analyzer_flakeref,
+                        attr_path: analysis_attr_path,
+                    }
+                    .into(),
+                ),
+                ..Default::default()
+            },
+            ..Default::default()
+        };
+
+        eval_analysis_command
+            .run_json(&nix, &Default::default())
+            .map_err(|nix_error| {
+                PublishError::DrvMetadata(
+                    self.attr_path.clone(),
+                    self.publish_ref.clone(),
+                    nix_error,
+                )
+            })
+            .await
+    }
+
+    /// Resolve the metadata of the flake holding the published package
+    async fn get_flake_metadata(&self) -> PublishResult<FlakeMetadata> {
+        let nix: NixCommandLine = self.flox.nix(Default::default());
+
+        let locked_ref_command = runix::command::FlakeMetadata {
+            flake_ref: Some(self.publish_ref.clone().into_inner().into()),
+            ..Default::default()
+        };
+
+        locked_ref_command
+            .run_typed(&nix, &Default::default())
+            .map_err(|nix_err| PublishError::FlakeMetadata(self.publish_ref.clone(), nix_err))
+            .await
+    }
+}
+
+impl<'flox> Publish<'flox, NixAnalysis> {
+    /// copy the outputs and dependencies of the package to binary store
+    pub async fn upload_binary(self) -> PublishResult<Publish<'flox, NixAnalysis>> {
+        todo!()
+    }
+
+    #[allow(unused)] // until implemented
+    async fn get_binary_cache_metadata(
+        &self,
+        substituter: SubstituterUrl,
+    ) -> PublishResult<CacheMeta> {
+        todo!()
+    }
+
+    /// write snapshot to catalog and push to origin
+    pub async fn push_catalog(self) -> PublishResult<()> {
```

Comment:
I think loading the catalog, writing to the catalog, and pushing the catalog are three separate operations. Splitting them out would make them easier to test.

Reply:
yes it would
understand this method more as a draft to be completed in a separate pr